### PR TITLE
Auth/PM-7321 - Registration with Email Verification - Registration Finish Component Implementation

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## 🎟️ Tracking

[PM-7321](https://bitwarden.atlassian.net/browse/PM-7321)
Associated Server PR: https://github.com/bitwarden/server/pull/4182

## 📔 Objective

To implement the registration finish component. 

### Tools & Admin Console Team code changes:
I had to replace the static `registerRoute` setting logic with a reactive call to the new `registerRouteService` to determine the proper register route in several of your components.

### Known defects:
The MP strength meter is unfortunately implemented using Bootstrap which isn't available on desktop or browser so... we will have to fix that in a separate PR. 

## 📸 Screenshots

Note: email verification can be disabled via a self hosted setting (separate from the feature flag). 

### Web

Email verification scenario:

https://github.com/bitwarden/clients/assets/116684653/77da4937-c1a3-4374-8f8e-84adf1baab58

Org Invite Scenario:

https://github.com/bitwarden/clients/assets/116684653/863106d5-1334-4e17-ad68-23d3e1e3488f

Registration with Email Verification Disabled  - Successful Account Creation:

https://github.com/bitwarden/clients/assets/116684653/9460a658-546c-407c-893e-3cffaff4aad1

New validation in action to prevent invalid tokens from being accepted:

https://github.com/bitwarden/clients/assets/116684653/91145c7f-bfaa-4111-9366-456badce4da4

### Desktop

Registration with Email Verification Enabled - Successful Account Creation

https://github.com/bitwarden/clients/assets/116684653/5f0d8b12-755a-4987-a51d-e2a859e9b871

Registration with Email Verification Disabled - Successful Account Creation

https://github.com/bitwarden/clients/assets/116684653/7f7ab4c0-d0a9-4129-b150-5ab8331200db

### Browser Extension

Registration with Email Verification Enabled - Successful Account Creation

https://github.com/bitwarden/clients/assets/116684653/4fa8057a-41da-4e75-9b59-fc767e31c452

Registration with Email Verification Disabled - Successful Account Creation

https://github.com/bitwarden/clients/assets/116684653/4888d414-e47c-4801-9fd1-24a114039d88

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-7321]: https://bitwarden.atlassian.net/browse/PM-7321?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/web/src/app/tools/send/access.component.ts [hunk 1]
```diff
@@ -56,7 +56,7 @@ export class AccessComponent implements OnInit {
   protected formGroup = this.formBuilder.group({});
 
   // TODO: remove when email verification flag is removed
-  registerRoute = "/register";
+  registerRoute$ = this.registerRouteService.registerRoute$();
```
Comment (audreyality) [inline apps/web/src/app/tools/send/access.component.ts:59]: 👍🏻 Love to see this coming from a service! <3

**Review by audreyality — DISMISSED** [review]

**Review by eliykat — APPROVED** [review]

File: apps/browser/src/_locales/en/messages.json [hunk 2]
```diff
@@ -556,6 +556,18 @@
   "security": {
     "message": "Security"
   },
+  "confirmMasterPassword": {
+    "message": "Confirm master password"
+  },
+  "masterPassword": {
```
Comment (jlf0dev) [inline apps/browser/src/_locales/en/messages.json:562]: nit: we have an existing one under `masterPass`, i'm sure we had the same thought process 😆

File: apps/web/src/app/auth/core/services/registration/web-registration-finish.service.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    const registerRequest = new RegisterFinishRequest(
+      email,
+      emailVerificationToken,
+      passwordInputResult.masterKeyHash,
+      passwordInputResult.hint,
+      encryptedUserKey,
+      userAsymmetricKeysRequest,
+      passwordInputResult.kdfConfig.kdfType,
+      passwordInputResult.kdfConfig.iterations,
+    );
```
Comment (jlf0dev) [inline apps/web/src/app/auth/core/services/registration/web-registration-finish.service.ts:89]: suggestion:
```suggestion
    const registerRequest = super.buildRegisterRequest(email, emailVerificationToken, passwordInputResult, encryptedUserKey, userAsymmetricKeys)
```

**Review by jlf0dev — APPROVED** [review]
Great work! Had some minor suggestions but nothing blocking

File: apps/web/src/app/auth/core/services/registration/web-registration-finish.service.ts (hunk 3 above)
Comment (JaredSnider-Bitwarden) [reply apps/web/src/app/auth/core/services/registration/web-registration-finish.service.ts]: Great call! Fixed!

File: apps/browser/src/_locales/en/messages.json (hunk 2 above)
Comment (JaredSnider-Bitwarden) [reply apps/browser/src/_locales/en/messages.json:562]: Can't really resolve this b/c the single libs/auth component approach requires the same translation keys across clients... unless we conditioned on client type which would be not great.

**Review by jlf0dev — APPROVED** [review]

**Review by audreyality — APPROVED** [review]

**Review by eliykat — APPROVED** [review]

**Review by coroiu — APPROVED** [review]
Only reviewed files owned by Platform 👍

**Pull request merged**